### PR TITLE
Module nccl

### PR DESCRIPTION
**Pull request by rfj82982**
Separation of the NCCL code from decomp2d module. This has a first implementation of the transpose for complex using NCCL. This PR should be included after the cuMPI (#160). Further work is on going to improve the transpose operation for complex.

### CODE REVIEW

File: src/decomp_2d_nccl.f90 [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_BCAST")
+
+      nccl_stat = ncclCommInitRank(nccl_comm_2decomp, nproc, nccl_uid_2decomp, nrank)
+      if (nccl_stat /= ncclSuccess) call decomp_2d_abort(__FILE__, __LINE__, nccl_stat, "ncclCommInitRank")
+      cuda_stat = cudaStreamCreate(cuda_stream_2decomp)
+      if (cuda_stat /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "cudaStreamCreate")
+   
+   end subroutine decomp_2d_nccl_init
+   !
+   ! init of the arrays
```
Comment (mathrack) [inline src/decomp_2d_nccl.f90:104]: ?

File: src/decomp_2d_init_fin.f90 [hunk 2]
```diff
@@ -177,35 +155,36 @@ subroutine decomp_2d_init_ref(nx, ny, nz, p_row, p_col, periodic_bc, comm)
...
+     !if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLREDUCE")
+
+     !if (nrank == 0) then
+     !   nccl_stat = ncclGetUniqueId(nccl_uid_2decomp)
+     !end if
+     !call MPI_Bcast(nccl_uid_2decomp, int(sizeof(ncclUniqueId)), MPI_BYTE, 0, decomp_2d_comm, ierror)
+     !if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_BCAST")
+
+     !nccl_stat = ncclCommInitRank(nccl_comm_2decomp, nproc, nccl_uid_2decomp, nrank)
+     !cuda_stat = cudaStreamCreate(cuda_stream_2decomp)
```
Comment (pbartholomew08) [inline src/decomp_2d_init_fin.f90:187]: Delete?

File: src/decomp_2d_init_fin.f90 [hunk 3]
```diff
@@ -232,34 +211,33 @@ end subroutine decomp_2d_init_ref
   subroutine decomp_2d_finalize_ref
 
      implicit none
-#if defined(_GPU) && defined(_NCCL)
-     type(ncclResult) :: nccl_stat
-#endif
+!#if defined(_GPU) && defined(_NCCL)
+!     type(ncclResult) :: nccl_stat
+!#endif
```
Comment (pbartholomew08) [inline src/decomp_2d_init_fin.f90:216]: Delete?

File: src/decomp_2d_init_fin.f90 [hunk 4]
```diff
@@ -232,34 +211,33 @@ end subroutine decomp_2d_init_ref
...
      decomp_buf_size = 0
      deallocate (work1_r, work2_r, work1_c, work2_c)
 #if defined(_GPU)
-     deallocate (work1_r_d, work2_r_d, work1_c_d, work2_c_d)
-
+     call decomp_2d_cumpi_fin()
 #if defined(_NCCL)
-     nccl_stat = ncclCommDestroy(nccl_comm_2decomp)
+     call decomp_2d_nccl_fin()
+     !nccl_stat = ncclCommDestroy(nccl_comm_2decomp)
```
Comment (pbartholomew08) [inline src/decomp_2d_init_fin.f90:236]: Delete?

File: src/transpose_y_to_x.f90 [hunk 5]
```diff
@@ -30,10 +30,10 @@ subroutine transpose_y_to_x_real(src, dst, decomp)
...
      TYPE(DECOMP_INFO), intent(IN) :: decomp
 
-#if defined(_GPU) && defined(_NCCL)
-     type(ncclResult) :: nccl_stat
-     integer :: col_rank_id, cuda_stat
-#endif
+!#if defined(_GPU) && defined(_NCCL)
+!     type(ncclResult) :: nccl_stat
+!     integer :: col_rank_id, cuda_stat
+!#endif
```
Comment (pbartholomew08) [inline src/transpose_y_to_x.f90:36]: Delete?

**Comment by pbartholomew08** [comment]
I'm assuming #160 should be merged first as this has changes that are defined there (I think)?

**Comment by pbartholomew08** [comment]
Marking as draft as depends on  #160

File: examples/fft_physical_x/fft_c2c_x.f90 [hunk 6]
```diff
@@ -129,12 +129,12 @@ program fft_c2c_x
    ierror = cudaDeviceSynchronize()
 #endif
 
-   call MPI_ALLREDUCE(t2, t1, 1, real_type, MPI_SUM, &
+   call MPI_ALLREDUCE(t2, t1, 1, MPI_DOUBLE, MPI_SUM, &
```
Comment (pbartholomew08) [inline examples/fft_physical_x/fft_c2c_x.f90:132]: To be absolutely safe this should be `MPI_DOUBLE_PRECISION` I think, and similar below and in other fft examples

**Review by pbartholomew08 — CHANGES_REQUESTED** [review]

File: src/decomp_2d_nccl.f90 (hunk 1 above)
Comment (rfj82982) [reply src/decomp_2d_nccl.f90:105]: In this a question on the comment?

File: src/decomp_2d_init_fin.f90 (hunk 4 above)
Comment (rfj82982) [reply src/decomp_2d_init_fin.f90]: We can change the name, not really an issue for me. Don’t remember why i chose destroy over another name

Comment (pbartholomew08) [reply src/decomp_2d_init_fin.f90]: I meant the commented out line, not a comment on the name :)

File: src/decomp_2d.f90 [hunk 7]
```diff
@@ -80,11 +81,11 @@ module decomp_2d
    integer(kind(D2D_DEBUG_LEVEL_OFF)), public, save :: decomp_debug = D2D_DEBUG_LEVEL_OFF
 #endif
 
-#if defined(_GPU)
-#if defined(_NCCL)
-   integer, save :: row_rank, col_rank
-#endif
-#endif
+!#if defined(_GPU)
```
Comment (mathrack) [inline src/decomp_2d.f90:84]: Clean the commented code ?

File: src/decomp_2d.f90 [hunk 8]
```diff
@@ -132,16 +133,16 @@ module decomp_2d
...
-
-#if defined(_NCCL)
-   integer col_comm_size, row_comm_size
-   integer, allocatable, dimension(:) :: local_to_global_col, local_to_global_row
-   type(ncclUniqueId) :: nccl_uid_2decomp
-   type(ncclComm) :: nccl_comm_2decomp
-   integer(kind=cuda_stream_kind) :: cuda_stream_2decomp
-#endif
-#endif
+!#if defined(_GPU)
```
Comment (mathrack) [inline src/decomp_2d.f90:136]: Clean the commented code ?

File: src/decomp_2d_nccl.f90 (hunk 1 above)
Comment (mathrack) [reply src/decomp_2d_nccl.f90:105]: The comment for decomp_2d_ccl_fin should not be "init of the arrays"

File: src/decomp_2d_nccl.f90 [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+   use decomp_2d_mpi
+   use decomp_2d_cumpi
+   use nccl
+
+   implicit none
+   
+   private        ! Make everything private unless declared public
+
+#ifdef DOUBLE_PREC
+   type(ncclDataType),public :: ncclType = ncclDouble
```
Comment (mathrack) [inline src/decomp_2d_nccl.f90:28]: Add the parameter attribute ?

File: src/decomp_2d_nccl.f90 [hunk 10]
```diff
@@ -0,0 +1,310 @@
+!=======================================================================
```
Comment (mathrack) [inline src/decomp_2d_nccl.f90:1]: Apply fprettify to the file ?

File: src/decomp_2d_nccl.f90 [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+   !
+   ! init of the arrays
+   ! 
+   subroutine decomp_2d_nccl_fin()
+      
+      implicit none
+      
+      type(ncclResult) :: nccl_stat
+
+      nccl_stat = ncclCommDestroy(nccl_comm_2decomp)
```
Comment (mathrack) [inline src/decomp_2d_nccl.f90:113]: Check the error ?

**Review by mathrack — CHANGES_REQUESTED** [review]

File: src/decomp_2d.f90 (hunk 7 above)
Comment (rfj82982) [reply src/decomp_2d.f90:84]: Code has been cleaned

File: src/decomp_2d_nccl.f90 (hunk 9 above)
Comment (rfj82982) [reply src/decomp_2d_nccl.f90:28]: yes

File: src/decomp_2d_nccl.f90 (hunk 11 above)
Comment (rfj82982) [reply src/decomp_2d_nccl.f90:113]: Done

File: src/decomp_2d_nccl.f90 (hunk 10 above)
Comment (mathrack) [reply src/decomp_2d_nccl.f90:1]: Done

**Review by mathrack — APPROVED** [review]

**Review by pbartholomew08 — APPROVED** [review]

**Pull request merged**